bitcoin/src/taproot.rs

```diff
@@ -724,25 +724,43 @@ impl TryFrom<NodeInfo> for TapTree {
     }
 }
 
-/// Iterator for a taproot script tree, operating in DFS order over leaf depth and
-/// leaf script pairs.
-/// This is guaranteed to not contain any hidden nodes.
+/// Iterator for a taproot script tree, operating in DFS order yielding [`ScriptLeaf`].
+///
+/// Returned by [`TapTree::script_leaves`]. [`TapTree`] does not allow hidden nodes,
+/// so this iterator is guaranteed to yield all known leaves.
+pub struct ScriptLeaves<'tree> {
+    leaf_iter: LeafNodes<'tree>,
+}
 
-// TODO: Enforce this in type system in a later PR. Users should not need to unwrap
-// as script leaf when there are no hidden nodes.
-// An idea can be to have the iterator return a tuple (depth, script, version).
-pub struct TapTreeIter<'tree> {
-    leaf_iter: core::slice::Iter<'tree, ScriptLeaf>,
+impl<'tree> Iterator for ScriptLeaves<'tree> {
+    type Item = ScriptLeaf<'tree>;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+            ScriptLeaf::from_leaf_node(self.leaf_iter.next()?)
+    }
 }
 
-impl<'tree> Iterator for TapTreeIter<'tree> {
-    type Item = &'tree ScriptLeaf;
+impl<'tree> ExactSizeIterator for ScriptLeaves<'tree> {
+    #[inline]
+    fn len(&self) -> usize { self.leaf_iter.len() }
+}
+
+/// Iterator for a taproot script tree, operating in DFS order yielding [`LeafNode`].
+///
+/// Returned by [`NodeInfo::leaf_nodes`]. This can potentially yield hidden nodes.
+pub struct LeafNodes<'a> {
+    leaf_iter: core::slice::Iter<'a, LeafNode>,
+}
+
+impl<'a> Iterator for LeafNodes<'a> {
+    type Item = &'a LeafNode;
 
     #[inline]
     fn next(&mut self) -> Option<Self::Item> { self.leaf_iter.next() }
 }
 
-impl<'tree> ExactSizeIterator for TapTreeIter<'tree> {
+impl<'tree> ExactSizeIterator for LeafNodes<'tree> {
```

Comment:
Other traits for consideration here and above: `FusedIterator`, `DoubleEndedIterator`